src/modules/retransmitter/retransmitter.cpp

```diff
@@ -132,36 +142,40 @@ extern "C" void task_retransmitter_fn(void) {
     ros_echronos::Publisher<std_msgs::Float64> back_right_s(topics[6], (std_msgs::Float64*)msg_buf_back_right_s, BUF_SIZE, false);
     publishers[7] = &back_right_s;
 
-
     for (size_t i = 0; i < NUM_MSG; i++) {
         publishers[i]->init(nh);
     }
-    
-    std_msgs::Float64 msg;
-    struct messageAdapter serial;
-    int counter = 0;
+
     while(true) {
-        wait_for_msg();
 
+        wait_for_msg();
         for (size_t i = 0; i < sizeof(struct message); i++) {
-            serial.data.structBytes[i] = (uint8_t)UARTgetc();
+            serial.data.structBytes[i] = (uint8_t) UARTgetc();
+            buf_reading.structBytes[i] = serial.data.structBytes[i];
         }
+        buf_ready = buf_reading;
+        is_buffer_ready = true;
+        UARTwrite((const char*)startMagic, 4);
+    }
+}
 
-        /*if (serial.data.msg.endMagic != endMagic) {
-            continue;
-        }*/
-
+extern "C" void task_publish_buffer_fn(void) {
+    ros_echronos::NodeHandle nh = *shared_nh;
+    std_msgs::Float64 msg;
 
+    while(true) {
+        // wait for buffer to fill up
+        while (!is_buffer_ready);
+        buf_sending = buf_ready;

```

Comment:
you probably need a mutex round this line (and the assignment to it on ln156). Assignment of structs isn't guaranteed to be threadsafe and you could end up with corupt data

Reply:
I don't think a mutex is necessary here. The loop in line 169:
```C
while (!is_buffer_ready);
```
waits for reading to be complete.
What I suspect might be problematic is that I am just making `buf_sending` point to `buf_ready` (as these are pointers). So if the contents of `buf_ready` change later, the contents of `buf_sending` would change too. Is there a way to make a copy of the contents of the struct?

Reply:
Actually, looking back at it, these are not pointers, so the assignment operator should just make a copy of the struct.

Reply:
Added mutexes